File: python/1020.Number-of-Enclaves.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        visited = set()
        land = 0
        border_land = 0

        def dfs(r, c):
            if r < 0 or c < 0 or r >= rows or c >= cols or grid[r][c] == 0 or (r, c) in visited:
                return 0
            visited.add((r, c))
            count = 1 + dfs(r + 1, c) + dfs(r - 1, c) + dfs(r, c + 1) + dfs(r, c - 1)
            return count

        for r in range(rows):
            for c in range(cols):
                land += grid[r][c]

                if grid[r][c] == 1 and (r in [0, rows - 1] or c in [0, cols - 1]) and (r, c) not in visited:
                    border_land += dfs(r, c)

        return land - border_land

File: javascript/1020.Number-of-Enclaves.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var numEnclaves = function (grid) {
  const rows = grid.length;
  const cols = grid[0].length;
  const visited = new Set();
  let land = 0;
  let border_land = 0;

  const dfs = (r, c) => {
    if (
      r < 0 ||
      c < 0 ||
      r >= rows ||
      c >= cols ||
      grid[r][c] === 0 ||
      visited.has(`${r}-${c}`)
    ) {
      return 0;
    }

    visited.add(`${r}-${c}`);
    const count =
      1 + dfs(r + 1, c) + dfs(r - 1, c) + dfs(r, c + 1) + dfs(r, c - 1);

    return count;
  };

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      land += grid[r][c];

      if (
        grid[r][c] === 1 &&
        ([0, rows - 1].includes(r) || [0, cols - 1].includes(c)) &&
        !visited.has(`${r}-${c}`)
      ) {
        border_land += dfs(r, c);
      }
    }
  }

  return land - border_land;
};
